Feature/#296-implement-actors-screen

fatmahgazy opened the pull request:
**feat: implement people UI screen with navigation, state management,room db and pagination**
- 

- Implement People screen UI

- Set up navigation with nested graph structure

- Define and manage PeopleUiState

- Handle user interactions via a listener interface

- Connect ViewModel to domain layer using clean architecture

- Use remote data source (Ktor) and local cache (Room)

- Map remote DTOs to domain models and local entities**

- Add pagination support with remote loading and local Room caching



https://github.com/user-attachments/assets/3314f767-e4db-4f57-952c-759368b411e2

fatmahgazy committed:
feat: impelemen people ui screen with navigation and ui state , mapper, interaction listener and handle use case and view model with remote data source and local using room.
implement pagination

fatmahgazy committed:
refactor: format ui screen

fatmahgazy committed:
Merge remote-tracking branch 'origin/develop' into feature/#296-implement-actors-screen

# Conflicts:
#	app/src/main/java/com/baghdad/novix/MainActivity.kt
#	design_system/src/main/java/com/baghdad/design_system/component/ActorCard.kt

fatmahgazy committed:
refactor: resolve conflict

Malak187 commented on app/src/main/java/com/baghdad/novix/MainActivity.kt:
You don't have to push this file into develop. We should keep the main screen as our entry screen

Malak187 commented on app/src/main/java/com/baghdad/novix/di/ViewModelModule.kt:
```diff
@@ -69,4 +70,7 @@ val viewModelModule = module {
     }
+    viewModel {
+        PeopleViewModel(get())
+    }
```
Why didn't you use viewmodelof as we always used to, since no parameters are needed here?

Hasa9nAli replied:
i agree with malak you can use  `viewModelOf(::CarsViewModel)` instead viewModel {
        PeopleViewModel(get())
    } . It more consistency

Malak187 commented on domain/src/main/java/com/baghdad/domain/repository/PeopleRepository.kt:
why did you create repository for people? we can use the actor repository we already have, and we also don't have to use the people term.  You can use actor

Malak187 commented on entity/src/main/java/com/baghdad/entity/person/People.kt:
You can depend on the actor entity. We don't have to create a new entity

Malak187 commented on remote_datasource/src/main/java/com/baghdad/remoteDataSource/RemotePeopleDataSourceImpl.kt:
We already have a function to handle paging. It is better to use it to apply pagination logic. You can take search flow as an example

Malak187 commented on remote_datasource/src/main/java/com/baghdad/remoteDataSource/response/people/PopularPeopleResponse.kt:
Please make all the variables nullable

Malak187 commented on repository/src/main/java/com/baghdad/repository/PeopleRepositoryImpl.kt:
```diff
@@ -0,0 +1,44 @@
+                }
+
+                remoteResult
+            },
```
Why do we need this?

Malak187 commented on repository/src/main/java/com/baghdad/repository/model/PeopleDto.kt:
You also don't have to create a new Dto , you can depend on the ActorDto we already have

Malak187 commented on ui/src/main/java/com/baghdad/ui/feature/people/PeopleScreen.kt:
```diff
@@ -0,0 +1,85 @@
+    viewModel: PeopleViewModel = koinViewModel(), handleNavigation: (PeopleNavEvent) -> Unit
+) {
+    val uiState by viewModel.uiState.collectAsStateWithLifecycle()
+
```
you have to collect the snackbar state here

Malak187 commented on ui/src/main/java/com/baghdad/ui/feature/people/PeopleScreen.kt:
```diff
@@ -0,0 +1,85 @@
+fun PeopleScreenContent(
+    uiState: PeopleUiState, listner: PeopleInteractionListener
+) {
+    val peopleItems = uiState.people.collectAsLazyPagingItems()
```
please move the collect to the PeopleScreen itself to apply state hoisting

Malak187 commented on ui/src/main/java/com/baghdad/ui/feature/people/PeopleScreen.kt:
```diff
@@ -0,0 +1,85 @@
+            onGoBackClick = listner::onBackClick,
+            screenTitle = stringResource(R.string.people),
+        ) {}
+
```
You need to apply scaffold here as we did in other screens

Malak187 commented on ui/src/main/java/com/baghdad/ui/feature/people/PeopleScreen.kt:
```diff
@@ -0,0 +1,85 @@
+            modifier = Modifier
+                .fillMaxSize()
+                .padding(horizontal = 16.dp)
+        ) {
```
You have to use LazyPagingColumn component

Malak187 commented on ui/src/main/java/com/baghdad/ui/navigation/NovixNavHost.kt:
```diff
@@ -39,5 +40,6 @@ fun NovixNavHost(
         movieDetailsNavGraph(navController)
         tvShowDetailsNavGraph(navController)
         reviewsNavGraph(navController)
+        peopleNavGraph(navController)
```
You don't have to create new graph for your screen. It is already defined in the home nav graph

Malak187 commented on ui/src/main/java/com/baghdad/ui/navigation/graph/people/PeopleNavEvent.kt:
You also need to remove this. It is already there

Malak187 commented on ui/src/main/java/com/baghdad/ui/navigation/graph/people/PeopleNavGraph.kt:
same here. remove this

Malak187 commented on ui/src/main/java/com/baghdad/ui/navigation/route/Graph.kt:
same

Malak187 commented on ui/src/main/java/com/baghdad/ui/navigation/route/PeopleRoute.kt:
same. remove anything related to the navigation package

Malak187 commented on viewmodel/src/main/java/com/baghdad/viewmodel/people/PeopleUiState.kt:
isPaging, endReached, currentPage, all of these variables are not used. You have to remove them

Malak187 reviewed (changes requested):
Great effort Fatma. You just need to check the comments
and please try to use the term (actor) not (people or person) to keep the consistency along the app 
I also tried to run the app and I found out that the pagination is not applied.

Hasa9nAli commented on domain/src/main/java/com/baghdad/domain/repository/PeopleRepository.kt:
```diff
@@ -0,0 +1,8 @@
+
+interface PeopleRepository {
+ suspend fun getPopularPeople(page: Int): PagedResult<People>
+}
```
there already exist ActorRepository, you and use it and add your functionality on it 

<img width="220" height="50" alt="Screenshot 2025-07-23 at 12 55 09 PM" src="https://github.com/user-attachments/assets/7378f31c-5a15-42fe-9242-3966b2720e1f" />

Hasa9nAli commented on domain/src/main/java/com/baghdad/domain/usecase/people/GetPeopleUseCase.kt:
```diff
@@ -0,0 +1,18 @@
+import com.baghdad.domain.repository.PeopleRepository
+import com.baghdad.entity.person.People
+
+class GetPopularPeopleUseCase(
```
you can use GetPopularActorUseCase , People indicate to  global named

Hasa9nAli commented on local_datasource/src/main/java/com/baghdad/local_datasource/LocalPopularPeopleDataSourceImpl.kt:
```diff
@@ -0,0 +1,40 @@
+         executeWithErrorHandling(logger = logger){
+            peopleDao.deleteAllPopularPeople()
+        }
+    }
```
named same

Hasa9nAli commented on local_datasource/src/main/java/com/baghdad/local_datasource/roomDB/dao/PeopleDao.kt:
```diff
@@ -0,0 +1,31 @@
+import com.baghdad.local_datasource.roomDB.entity.PopularPeopleEntity
+
+@Dao
+interface PeopleDao {
```
same

Hasa9nAli commented on local_datasource/src/main/java/com/baghdad/local_datasource/roomDB/dao/PeopleDao.kt:
```diff
@@ -0,0 +1,31 @@
+@Dao
+interface PeopleDao {
+    @Upsert
+    suspend fun upsertPopularPeople(people: List<PopularPeopleEntity>)
```
same

Hasa9nAli commented on local_datasource/src/main/java/com/baghdad/local_datasource/roomDB/dao/PeopleDao.kt:
```diff
@@ -0,0 +1,31 @@
+    LIMIT :pageSize OFFSET :offset
+    """
+    )
+    suspend fun getAllPopularPeople(
```
same

Hasa9nAli commented on local_datasource/src/main/java/com/baghdad/local_datasource/roomDB/dao/PeopleDao.kt:
```diff
@@ -0,0 +1,31 @@
+    suspend fun getAllPopularPeople(
+        pageSize: Int,
+        offset: Int
+    ): List<PopularPeopleEntity>
```
same

Hasa9nAli reviewed (commented):
good work fatma , please check comment

fatmahgazy committed:
refactor: delete old data from remote  function and add prams to fix paganation

fatmahgazy committed:
refactor: add view model of instead view model and koin will provide the dependancies it self

fatmahgazy committed:
refactor: add view model of instead view model and koin will provide the dependancies it self
make repo and remote data siurce and local data source in the same class of actors and also dtos and edit translate to arabic then refactor name of people screen

fatmahgazy committed:
refactor: edit main activity to use home graph

MuhammedEdrees commented on app/src/main/java/com/baghdad/novix/MainActivity.kt:
```diff
@@ -4,16 +4,28 @@ import android.os.Bundle
+                    navController = navController,
+                    startDestination = Graph.HomeGraph
+                )
+
```
return this to be `MainScreen` instead.

fatmahgazy committed:
refactor: replace popular keyword to trending actor

fatmahgazy committed:
refactor: try to push the get actor by id funcation

fatmahgazy committed:
Merge remote-tracking branch 'origin/develop' into feature/#296-implement-actors-screen

# Conflicts:
#	remote_datasource/src/main/java/com/baghdad/remoteDataSource/RemoteActorDataSourceImpl.kt

fatmahgazy committed:
refactor: convert from ktor to retrofit

Hasa9nAli commented:
you have some problem please solve it CI not accept there are something not successful

MuhammedEdrees commented on ui/src/main/java/com/baghdad/ui/navigation/graph/actorDetails/ActorDetailsNavGraph.kt:
```diff
@@ -75,5 +75,11 @@ private fun handleActorDetailsNavigation(
+
+        is ActorDetailsNavEvent.NavigateToActorDetails -> navController.navigate(
+            ActorDetailsGraph(event.actorId)
+        )
```
Remove these from here as this screen belongs to home nav graph

MuhammedEdrees commented on ui/src/main/java/com/baghdad/ui/navigation/route/PeopleRoute.kt:
Remove

MuhammedEdrees commented on ui/src/main/java/com/baghdad/ui/navigation/graph/home/HomeNavGraph.kt:
```diff
@@ -31,7 +32,9 @@ fun NavGraphBuilder.homeNavGraph(navController: NavController) {
             DummyScreen(title = "TV Shows Screen")
         }
         composable<HomeRoute.ActorsScreen> {
```
rename the route to be same as screen name

MuhammedEdrees commented on viewmodel/src/main/java/com/baghdad/viewmodel/people/TrendingActorsInteractionListener.kt:
```diff
@@ -0,0 +1,6 @@
+
+interface TrendingActorsInteractionListener{
+    fun onBackClick()
+    fun onPeopleClick(peopleId: Long)
```
change peopleId and onPeopleClick

MuhammedEdrees commented on viewmodel/src/main/java/com/baghdad/viewmodel/people/TrendingActorViewModel.kt:
```diff
@@ -0,0 +1,44 @@
+        return BaseSnackBarMessage.UnknownError
+    }
+
+    private fun getPopularPeople() {
```
rename

MuhammedEdrees commented on viewmodel/src/main/java/com/baghdad/viewmodel/people/TrendingActorViewModel.kt:
```diff
@@ -0,0 +1,44 @@
+import com.baghdad.viewmodel.errorStates.BaseSnackBarMessage
+
+class TrendingActorViewModel(
+    private val getPeopleUseCase: GetTrendingActorUseCase
```
rename

MuhammedEdrees commented on viewmodel/src/main/java/com/baghdad/viewmodel/people/TrendingActorViewModel.kt:
```diff
@@ -0,0 +1,44 @@
+import com.baghdad.viewmodel.base.BaseViewModel
+import com.baghdad.viewmodel.errorStates.BaseSnackBarMessage
+
+class TrendingActorViewModel(
```
*TrendingActors

MuhammedEdrees commented on viewmodel/src/main/java/com/baghdad/viewmodel/people/TrendingActorViewModel.kt:
```diff
@@ -0,0 +1,44 @@
+package com.baghdad.viewmodel.people
```
rename package

MuhammedEdrees commented on viewmodel/src/main/java/com/baghdad/viewmodel/people/TrendingActorUiState.kt:
```diff
@@ -0,0 +1,18 @@
+import kotlinx.coroutines.flow.emptyFlow
+
+data class TrendingActorUiState(
+    val trendingActor: Flow<PagingData<TrendingActor>> = emptyFlow(),
```
*Actors

MuhammedEdrees commented on viewmodel/src/main/java/com/baghdad/viewmodel/people/TrendingActorUiEffect.kt:
```diff
@@ -0,0 +1,8 @@
+
+import com.baghdad.viewmodel.base.BaseUiEffect
+
+sealed class TrendingActorUiEffect :BaseUiEffect{
```
*Actors

MuhammedEdrees commented on ui/src/main/java/com/baghdad/ui/feature/trendingActors/TrendingActorsScreen.kt:
```diff
@@ -0,0 +1,117 @@
+                        ActorCard(
+                            actorName = actor.name,
+                            actorImage = actor.profilePictureURL,
+                            onClick = { listner.onPeopleClick(actor.id) },
```
Use `LazyPagingColumn` instead. It handles loading and error sttates well

MuhammedEdrees commented on ui/src/main/java/com/baghdad/ui/feature/trendingActors/TrendingActorsScreen.kt:
```diff
@@ -0,0 +1,117 @@
+                .background(Theme.color.surface)
+                .statusBarsPadding()
+        ) {
+            TopAppBar(
+                modifier = Modifier.padding(top = 12.dp),
+                onGoBackClick = listner::onBackClick,
+                screenTitle = stringResource(R.string.actors),
+            ) {}
```
pass top bar to the scaffold at topBar not inside the content

MuhammedEdrees commented on ui/src/main/java/com/baghdad/ui/feature/trendingActors/TrendingActorsScreen.kt:
```diff
@@ -0,0 +1,117 @@
+    listner: TrendingActorsInteractionListener,
+    snackBarState: SnackBarState
+) {
+    val trendingActors = uiState.trendingActor.collectAsLazyPagingItems()
+
```
Move this to the screen itself and pass it to the content

MuhammedEdrees commented on repository/src/main/java/com/baghdad/repository/datasource/local/LocalActorDataSource.kt:
```diff
@@ -12,4 +12,10 @@ interface LocalActorDataSource {
     suspend fun updateActor(actor: ActorDto)
     suspend fun searchActorsByName(name: String, page: Int, pageSize: Int = 20): List<ActorDto>
+    suspend fun addTrendingActor(people: List<ActorDto>)
+    suspend fun getTrendingActor(page: Int, pageSize: Int = 20): List<ActorDto>
```
*Actors

MuhammedEdrees commented on repository/src/main/java/com/baghdad/repository/datasource/local/LocalActorDataSource.kt:
```diff
@@ -12,4 +12,10 @@ interface LocalActorDataSource {
+    suspend fun addTrendingActor(people: List<ActorDto>)
+    suspend fun getTrendingActor(page: Int, pageSize: Int = 20): List<ActorDto>
+    suspend fun getTrendingActorById(id: Long): ActorDto
+    suspend fun deleteTrendingActor()
```
*Actors

MuhammedEdrees commented on repository/src/main/java/com/baghdad/repository/datasource/local/LocalActorDataSource.kt:
```diff
@@ -12,4 +12,10 @@ interface LocalActorDataSource {
     suspend fun deleteAllActors()
     suspend fun updateActor(actor: ActorDto)
     suspend fun searchActorsByName(name: String, page: Int, pageSize: Int = 20): List<ActorDto>
+    suspend fun addTrendingActor(people: List<ActorDto>)
```
*Actors also rename the parameter from people to actors

MuhammedEdrees commented on repository/src/main/java/com/baghdad/repository/ActorRepositoryImpl.kt:
```diff
@@ -38,5 +44,21 @@ class ActorRepositoryImpl(
         }
     }
 
+    override suspend fun getTrendingActor(page: Int): PagedResult<Actor> {
```
*Actors

MuhammedEdrees commented on repository/src/main/java/com/baghdad/repository/ActorRepositoryImpl.kt:
```diff
@@ -1,15 +1,21 @@
 
 class ActorRepositoryImpl(
     private val remoteActorDataSource: RemoteActorDataSource,
+    private val localPopularPeopleDataSource: LocalActorDataSource
```
localPopularPeopleDataSource -> localActorDataSource

MuhammedEdrees commented on local_datasource/src/main/java/com/baghdad/local_datasource/roomDB/dao/TrendingActorDao.kt:
```diff
@@ -0,0 +1,31 @@
+@Dao
+interface TrendingActorDao {
+    @Upsert
+    suspend fun upsertTrendingActors(people: List<TrendingActorEntity>)
```
people -> actors

MuhammedEdrees commented on local_datasource/src/main/java/com/baghdad/local_datasource/LocalActorDataSourceImpl.kt:
```diff
@@ -46,20 +48,46 @@ class LocalActorDataSourceImpl(
+            trendingActorDao.getTrendingActorById(id).toDto()
         }
+
+    override suspend fun deleteTrendingActor() = executeWithErrorHandling(logger = logger) {
```
*Actors

MuhammedEdrees commented on local_datasource/src/main/java/com/baghdad/local_datasource/roomDB/dao/TrendingActorDao.kt:
```diff
@@ -0,0 +1,31 @@
+    suspend fun getTrendingActorById(personId: Long): TrendingActorEntity
+
+    @Query("DELETE FROM trendingActor")
+    suspend fun deleteAllTrendingActor()
```
*Actors

MuhammedEdrees commented on domain/src/main/java/com/baghdad/domain/usecase/actor/GetTrendingActorUseCase.kt:
```diff
@@ -0,0 +1,15 @@
+import com.baghdad.domain.repository.ActorRepository
+import com.baghdad.entity.person.Actor
+
+class GetTrendingActorUseCase(
```
*Actors

MuhammedEdrees commented on domain/src/main/java/com/baghdad/domain/repository/ActorRepository.kt:
```diff
@@ -9,4 +10,5 @@ interface ActorRepository {
     suspend fun getActorMovies(actorId: Long): List<Movie>
     suspend fun getActorTvShows(actorId: Long): List<TvShow>
     suspend fun getActorGallery(actorId: Long): List<String>
+    suspend fun getTrendingActor(page: Int): PagedResult<Actor>
```
*Actors

MuhammedEdrees reviewed (changes requested):
Well done fatma, There is some inconsistencies in naming and some old undeleted code please delete it and you're good to go.

MuhammedEdrees commented:
Is the screen working for you? When i run the screen no data is returned and after logging this is the exception that occurs. 
<img width="800" height="56" alt="image" src="https://github.com/user-attachments/assets/e6ed4a34-340f-43d9-9f03-0d72db99245f" />

fatmahgazy committed:
refactor: add main screen to main activity

fatmahgazy committed:
refactor: rename people to actors and handle navigation

fatmahgazy committed:
refactor: fix error in unit test

FaresM0hamed commented on remote_datasource/src/main/java/com/baghdad/remoteDataSource/response/actor/PopularActorResponse.kt:
```diff
@@ -0,0 +1,25 @@
+package com.baghdad.remoteDataSource.response.actor
+
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class TrendingActorResponse(
+    val page: Int? = null,
+    val results: List<TrendingActorDetails>? = null,
+    @SerialName("total_pages") val totalPages: Int? = null,
+    @SerialName("total_results") val totalResults: Int? = null
+)
+
+@Serializable
+data class TrendingActorDetails(
+    @SerialName("adult") val adult: Boolean? = null,
+    @SerialName("id") val id: Int? = null,
+    @SerialName("name") val name: String? = null,
+    @SerialName("original_name") val originalName: String? = null,
+    @SerialName("media_type") val mediaType: String? = null,
+    @SerialName("popularity") val popularity: Double? = null,
+    @SerialName("gender") val gender: Int? = null,
+    @SerialName("known_for_department") val knownForDepartment: String? = null,
+    @SerialName("profile_path") val profilePath: String? = null
+)
```
Based on the code and the recent migration from Ktor to Retrofit, I expect a failure when fetching this data due to the use of @SerialName. Retrofit uses Gson , and @SerialName is specific to Kotlinx Serialization.

To avoid this issue, please replace @SerialName with @SerializedName, as used throughout the app. You can refer to any existing model in our project for examples.

fatmahgazy committed:
refactor: convert serial name to serialized name

fatmahgazy committed:
refactor: add mock of local data source in repo unit test

FaresM0hamed commented on remote_datasource/src/main/java/com/baghdad/remoteDataSource/response/actor/TrendingActorResponse.kt:
```diff
@@ -0,0 +1,25 @@
+
+@Serializable
+data class TrendingActorResponse(
+    val page: Int? = null,
+    val results: List<TrendingActorDetails>? = null,
```
add annotation here also

fatmahgazy replied:
@FaresM0hamed  annotation of what?

MuhammedEdrees replied:
`@SerializedName` annotation to avoid issues with proguard

fatmahgazy replied:
Add this annotation to page and results fields too ?
@MuhammedEdrees

FaresM0hamed commented on remote_datasource/src/main/java/com/baghdad/remoteDataSource/response/actor/TrendingActorResponse.kt:
```diff
@@ -0,0 +1,25 @@
+import com.google.gson.annotations.SerializedName
+import kotlinx.serialization.Serializable
+
+@Serializable
```
remove every @Serializable

fatmahgazy committed:
refactor: remove serialztion annotation

fatmahgazy committed:
refactor: add serialization name to page and result fields

fatmahgazy committed:
refactor: add page query to api service

fatmahgazy committed:
refactor: add main screen in the main activity and graph to home screen

MuhammedEdrees reviewed (approved):
Good job
